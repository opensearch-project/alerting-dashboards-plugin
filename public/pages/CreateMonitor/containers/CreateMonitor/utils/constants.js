/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import {
  WHERE_FILTER_ALLOWED_TYPES,
  OPERATORS_MAP,
} from '../../../components/MonitorExpressions/expressions/utils/constants';
import { MONITOR_TYPE } from '../../../../../utils/constants';
import { SUPPORTED_DOC_LEVEL_QUERY_OPERATORS } from '../../../components/DocumentLevelMonitorQueries/utils/constants';

export const BUCKET_COUNT = 5;

export const MATCH_ALL_QUERY = JSON.stringify({ size: 0, query: { match_all: {} } }, null, 4);

export const FORMIK_INITIAL_WHERE_EXPRESSION_VALUES = {
  fieldName: [{ label: '', type: WHERE_FILTER_ALLOWED_TYPES[0] }], // This is an array because the EuiCombobox returns an array of {label: string, type: string} objects.
  operator: OPERATORS_MAP.IS.value,
  fieldValue: '',
  fieldRangeStart: undefined,
  fieldRangeEnd: undefined,
};

/** Sample delegate
 *  {
      order: 1,
      monitor_id: '{{m1}}',
    }
 */
export const DEFAULT_ASSOCIATED_MONITORS_VALUE = {
  sequence: {
    delegates: [],
  },
};

export const FORMIK_INITIAL_VALUES = {
  /* CONFIGURE MONITOR */
  name: '',
  disabled: false,
  frequency: 'interval',
  timezone: [],
  daily: 0,
  period: { interval: 1, unit: 'MINUTES' },
  weekly: { mon: false, tue: false, wed: false, thur: false, fri: false, sat: false, sun: false },
  monthly: { type: 'day', day: 1 },
  cronExpression: '0 */1 * * *',

  /* DEFINE MONITOR */
  monitor_type: MONITOR_TYPE.QUERY_LEVEL,
  searchType: 'graph',
  clusterNames: [],
  uri: {
    api_type: '',
    clusters: [],
    path: '',
    path_params: '',
    url: '',
  },
  index: [],
  timeField: '',
  query: MATCH_ALL_QUERY,
  queries: [],
  description: '',
  aggregationType: 'count',
  fieldName: [],
  aggregations: [],
  groupBy: [],
  groupByField: [{ label: '' }],
  overDocuments: 'all documents',
  groupedOverTop: 5,
  groupedOverFieldName: 'bytes',
  bucketValue: 1,
  bucketUnitOfTime: 'h', // m = minute, h = hour, d = day
  filters: [], // array of FORMIK_INITIAL_WHERE_EXPRESSION_VALUES
  detectorId: '',
  associatedMonitors: DEFAULT_ASSOCIATED_MONITORS_VALUE,
  associatedMonitorsList: [],
  associatedMonitorsEditor: '',
  preventVisualEditor: false,
};

export const FORMIK_INITIAL_AGG_VALUES = {
  aggregationType: 'count',
  fieldName: '',
};

export const FORMIK_INITIAL_DOCUMENT_LEVEL_QUERY_VALUES = {
  id: undefined,
  queryName: '',
  field: '',
  operator: SUPPORTED_DOC_LEVEL_QUERY_OPERATORS[0],
  query: '',
  tags: [],
};

// TODO DRAFT: Is this an appropriate default to display when defining as an extraction query?
export const DEFAULT_DOCUMENT_LEVEL_QUERY = JSON.stringify(
  {
    description: 'DESCRIPTION_TEXT',
    queries: [
      {
        name: 'QUERY_NAME',
        query: { match_all: {} },
        tags: ['TAG_TEXT'],
      },
    ],
  },
  null,
  4
);

export const DEFAULT_COMPOSITE_AGG_SIZE = 50;

export const METRIC_TOOLTIP_TEXT = 'Extracted statistics such as simple calculations of data.';
export const TIME_RANGE_TOOLTIP_TEXT = 'The time frame of data the plugin should monitor.';
export const FILTERS_TOOLTIP_TEXT =
  'Use a filter to retrieve a subset of the original data source.';
export const GROUP_BY_TOOLTIP_TEXT = 'Specify a field whose every unique value can trigger alerts.';
