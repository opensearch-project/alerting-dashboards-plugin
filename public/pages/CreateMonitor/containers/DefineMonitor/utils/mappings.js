/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

export function shouldSkip(mapping) {
  const isDisabled = mapping.enabled === false;
  const hasIndexDisabled = mapping.index === false;
  const isNestedDataType = mapping.type === 'nested';
  return isDisabled || hasIndexDisabled || isNestedDataType;
}

export function resolvePath(path, field) {
  if (path) return `${path}.${field}`;
  return field;
}

export function getFieldsFromProperties(properties, dataTypes, path) {
  Object.entries(properties).forEach(([field, value]) => {
    getTypeFromMappings(value, dataTypes, resolvePath(path, field));
  });
}

export function getTypeFromMappings(mappings, dataTypes, path = '') {
  // Example structures of index mappings:
  // properties: { "field_name":{"properties": ...} }
  // properties: { "field_name":{"type": "text"} }
  if (shouldSkip(mappings)) return dataTypes;
  // if there are properties then type is inherently an object
  if (mappings.properties) {
    getFieldsFromProperties(mappings.properties, dataTypes, path);
    return dataTypes;
  }

  const type = mappings.type;

  if (dataTypes[type]) dataTypes[type].add(path);
  else dataTypes[type] = new Set([path]);
  return dataTypes;
}

export function getPathsPerDataType(mappings) {
  const dataTypes = {};
  Object.entries(mappings).forEach(([index, { mappings: docMappings }]) =>
    getTypeFromMappings(docMappings, dataTypes)
  );
  return dataTypes;
}
