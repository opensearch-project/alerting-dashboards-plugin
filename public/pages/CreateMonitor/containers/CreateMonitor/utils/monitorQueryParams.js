/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { SEARCH_TYPE } from '../../../../../utils/constants';

export const initializeFromQueryParams = (queryParams) => {
  return {
    searchType: queryParams.searchType || undefined,
    name: queryParams.name ? `${queryParams.name}-Monitor` : undefined,
    detectorId: queryParams.adId || undefined,
    period:
      queryParams.interval && queryParams.unit
        ? { interval: parseInt(queryParams.interval), unit: queryParams.unit }
        : undefined,
    adResultIndex: queryParams.adResultIndex || undefined,
  };
};
