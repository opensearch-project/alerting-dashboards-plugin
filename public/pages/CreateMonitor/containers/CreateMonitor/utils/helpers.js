/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import _ from 'lodash';
import queryString from 'query-string';
import { FORMIK_INITIAL_VALUES } from './constants';
import monitorToFormik from './monitorToFormik';
import { formikToMonitor } from './formikToMonitor';
import { MONITOR_TYPE } from '../../../../../utils/constants';
import { initializeFromQueryParams } from './monitorQueryParams';
import { backendErrorNotification, getDigitId } from '../../../../../utils/helpers';
import {
  formikToTrigger,
  formikToTriggerUiMetadata,
} from '../../../../CreateTrigger/containers/CreateTrigger/utils/formikToTrigger';
import { triggerToFormik } from '../../../../CreateTrigger/containers/CreateTrigger/utils/triggerToFormik';
import { TRIGGER_TYPE } from '../../../../CreateTrigger/containers/CreateTrigger/utils/constants';
import { FORMIK_INITIAL_AGG_VALUES } from '../../CreateMonitor/utils/constants';
import { getInitialTriggerValues } from '../../../../CreateTrigger/components/AddTriggerButton/utils';

export const getInitialValues = ({
  title,
  index,
  timeField,
  flyoutMode,
  location,
  monitorToEdit,
  edit,
  searchType,
  detectorId,
}) => {
  let initialValues = _.mergeWith(
    {},
    _.cloneDeep(FORMIK_INITIAL_VALUES),
    initializeFromQueryParams(queryString.parse(location.search)),
    (initialValue, queryValue) => (_.isEmpty(queryValue) ? initialValue : queryValue)
  );

  if (flyoutMode) {
    initialValues.name = `${title} ${getDigitId()}`;
    initialValues.index = index;
    initialValues.timeField = timeField;

    // Add trigger
    const monitorType = initialValues.monitor_type;
    const initialTrigger = getInitialTriggerValues({ flyoutMode, monitorType, triggers: [] });
    initialValues.triggerDefinitions = [initialTrigger];

    // Add aggregations
    initialValues.aggregations = [_.cloneDeep(FORMIK_INITIAL_AGG_VALUES)];

    if (searchType) {
      initialValues.searchType = searchType;
    }

    if (detectorId) {
      initialValues.detectorId = detectorId;
    }
  }

  if (edit && monitorToEdit) {
    const triggers = triggerToFormik(_.get(monitorToEdit, 'triggers', []), monitorToEdit);
    initialValues = {
      ...monitorToFormik(monitorToEdit),
      triggerDefinitions: triggers.triggerDefinitions,
    };
  }

  return initialValues;
};

export const getPlugins = async (httpClient) => {
  try {
    const pluginsResponse = await httpClient.get('../api/alerting/_plugins');
    if (pluginsResponse.ok) {
      return pluginsResponse.resp.map((plugin) => plugin.component);
    } else {
      console.error('There was a problem getting plugins list');
      return [];
    }
  } catch (e) {
    console.error('There was a problem getting plugins list', e);
    return [];
  }
};

export const prepareTriggers = ({
  trigger,
  triggerMetadata,
  monitor,
  edit,
  triggerToEdit = [],
}) => {
  const { ui_metadata: uiMetadata = {}, triggers, monitor_type } = monitor;
  let updatedTriggers;
  let updatedUiMetadata;

  if (edit) {
    updatedTriggers = _.isArray(trigger) ? trigger.concat(triggers) : [trigger].concat(triggers);
    updatedUiMetadata = {
      ...uiMetadata,
      triggers: { ...uiMetadata.triggers, ...triggerMetadata },
    };
  } else {
    let updatedTriggersMetadata = _.cloneDeep(uiMetadata.triggers || {});

    let triggerType;
    switch (monitor_type) {
      case MONITOR_TYPE.BUCKET_LEVEL:
        triggerType = TRIGGER_TYPE.BUCKET_LEVEL;
        break;
      case MONITOR_TYPE.DOC_LEVEL:
        triggerType = TRIGGER_TYPE.DOC_LEVEL;
        break;
      default:
        triggerType = TRIGGER_TYPE.QUERY_LEVEL;
        break;
    }

    if (_.isArray(triggerToEdit)) {
      const names = triggerToEdit.map((entry) => _.get(entry, `${triggerType}.name`));
      names.forEach((name) => delete updatedTriggersMetadata[name]);
      updatedTriggers = _.cloneDeep(trigger);
    } else {
      const { name } = _.get(triggerToEdit, `${triggerType}`);
      delete updatedTriggersMetadata[name];

      const findTriggerName = (element) => {
        return name === _.get(element, `${triggerType}.name`);
      };

      const indexToUpdate = _.findIndex(triggers, findTriggerName);
      updatedTriggers = triggers.slice();
      updatedTriggers.splice(indexToUpdate, 1, trigger);
    }

    updatedUiMetadata = {
      ...uiMetadata,
      triggers: { ...updatedTriggersMetadata, ...triggerMetadata },
    };
  }

  return { triggers: updatedTriggers, ui_metadata: updatedUiMetadata };
};

export const create = async ({
  monitor,
  formikBag,
  httpClient,
  notifications,
  history,
  onSuccess,
}) => {
  const { setSubmitting } = formikBag;

  try {
    const resp = await httpClient.post('../api/alerting/monitors', {
      body: JSON.stringify(monitor),
    });
    setSubmitting(false);
    const {
      ok,
      resp: { _id },
    } = resp;
    if (ok) {
      notifications.toasts.addSuccess(`Monitor "${monitor.name}" successfully created.`);
      history.push(`/monitors/${_id}`);

      if (onSuccess) {
        onSuccess({ monitorId: _id });
      }
    } else {
      console.log('Failed to create:', resp);
      backendErrorNotification(notifications, 'create', 'monitor', resp.resp);
    }
  } catch (err) {
    console.error(err);
    setSubmitting(false);
  }
};

export const update = async ({ history, updateMonitor, notifications, monitor, formikBag }) => {
  const { setSubmitting } = formikBag;
  const updatedMonitor = _.cloneDeep(monitor);
  try {
    const resp = await updateMonitor(updatedMonitor);
    setSubmitting(false);
    const { ok, id } = resp;
    if (ok) {
      notifications.toasts.addSuccess(`Monitor "${monitor.name}" successfully updated.`);
      history.push(`/monitors/${id}`);
    } else {
      console.log('Failed to update:', resp);
    }
  } catch (err) {
    console.error(err);
    setSubmitting(false);
  }
};

export const submit = ({
  values,
  formikBag,
  edit,
  triggerToEdit,
  history,
  updateMonitor,
  notifications,
  httpClient,
  onSuccess,
}) => {
  let monitor = formikToMonitor(values);

  if (!_.isEmpty(_.get(values, 'triggerDefinitions'))) {
    const monitorUiMetadata = _.get(monitor, 'ui_metadata', {});
    const triggerMetadata = formikToTriggerUiMetadata(values, monitorUiMetadata);
    const triggers = prepareTriggers({
      trigger: formikToTrigger(values, monitorUiMetadata),
      triggerMetadata,
      monitor,
      edit,
      triggerToEdit,
    });
    monitor = { ...monitor, ...triggers };
  }

  if (edit) {
    update({ history, updateMonitor, notifications, monitor, formikBag });
  } else {
    create({ history, monitor, formikBag, httpClient, notifications, onSuccess });
  }
};
