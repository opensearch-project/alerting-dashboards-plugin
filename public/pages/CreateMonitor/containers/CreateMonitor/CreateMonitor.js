/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React, { Component, Fragment } from 'react';
import _ from 'lodash';
import { FieldArray, Formik } from 'formik';
import {
  EuiButton,
  EuiButtonEmpty,
  EuiFlexGroup,
  EuiFlexItem,
  EuiSpacer,
  EuiTitle,
} from '@elastic/eui';
import DefineMonitor from '../DefineMonitor';
import { FORMIK_INITIAL_VALUES } from './utils/constants';
import { formikToMonitor } from './utils/formikToMonitor';
import { SEARCH_TYPE } from '../../../../utils/constants';
import { SubmitErrorHandler } from '../../../../utils/SubmitErrorHandler';
import MonitorDetails from '../MonitorDetails';
import ConfigureTriggers from '../../../CreateTrigger/containers/ConfigureTriggers';
import { triggerToFormik } from '../../../CreateTrigger/containers/CreateTrigger/utils/triggerToFormik';
import { getInitialValues, getPlugins, submit } from './utils/helpers';
import { createSavedObjectAssociation } from '../../../../utils/savedObjectHelper';

export default class CreateMonitor extends Component {
  static defaultProps = {
    edit: false,
    monitorToEdit: null,
    detectorId: null,
    updateMonitor: () => {},
    isDarkMode: false,
  };

  constructor(props) {
    super(props);

    const { location, edit, monitorToEdit } = props;
    const initialValues = getInitialValues({ location, monitorToEdit, edit });
    let triggerToEdit;

    if (edit && monitorToEdit) {
      triggerToEdit = triggerToFormik(_.get(monitorToEdit, 'triggers', []), monitorToEdit);
    }

    this.state = {
      plugins: [],
      response: null,
      performanceResponse: null,
      initialValues,
      triggerToEdit,
    };

    this.onCancel = this.onCancel.bind(this);
    this.onSubmit = this.onSubmit.bind(this);
  }

  componentDidMount() {
    const { httpClient } = this.props;

    const updatePlugins = async () => {
      const newPlugins = await getPlugins(httpClient);
      this.setState({ plugins: newPlugins });
    };

    updatePlugins();
    this.setSchedule();
  }

  resetResponse() {
    this.setState({ response: null, performanceResponse: null });
  }

  onCancel() {
    if (this.props.edit) this.props.history.goBack();
    else this.props.history.push('/monitors');
  }

  setSchedule = () => {
    const { edit, monitorToEdit } = this.props;
    const { initialValues } = this.state;

    if (edit) {
      const schedule = _.get(monitorToEdit, 'schedule', FORMIK_INITIAL_VALUES.period);
      const scheduleType = _.keys(schedule)[0];
      switch (scheduleType) {
        case 'cron':
          _.set(initialValues, 'frequency', 'cronExpression');
          break;
        default:
          _.set(initialValues, 'period', schedule.period);
          break;
      }
    }
  };

  onSubmit(values, formikBag) {
    const { edit, history, updateMonitor, notifications, httpClient } = this.props;
    const { triggerToEdit } = this.state;

    submit({
      values,
      formikBag,
      edit,
      triggerToEdit,
      history,
      updateMonitor,
      notifications,
      httpClient,
      onSuccess: async ({ monitor }) => {
        notifications.toasts.addSuccess(`Monitor "${monitor.name}" successfully created.`);
      },
    });
  }

  onCloseTrigger = () => {
    this.props.history.push({ ...this.props.location, search: '' });
  };

  componentWillUnmount() {
    this.props.setFlyout(null);
  }

  render() {
    const {
      edit,
      history,
      httpClient,
      location,
      monitorToEdit,
      notifications,
      isDarkMode,
      notificationService,
    } = this.props;
    const { initialValues, plugins } = this.state;

    return (
      <div style={{ padding: '25px 50px' }}>
        <Formik initialValues={initialValues} onSubmit={this.onSubmit} validateOnChange={false}>
          {({ values, errors, handleSubmit, isSubmitting, isValid, touched }) => (
            <Fragment>
              <EuiTitle size="l">
                <h1>{edit ? 'Edit' : 'Create'} monitor</h1>
              </EuiTitle>
              <EuiSpacer />

              <MonitorDetails
                values={values}
                errors={errors}
                history={history}
                httpClient={httpClient}
                monitorToEdit={monitorToEdit}
                plugins={plugins}
                isAd={values.searchType === SEARCH_TYPE.AD}
                detectorId={this.props.detectorId}
                setFlyout={this.props.setFlyout}
              />
              <EuiSpacer />

              {values.searchType !== SEARCH_TYPE.AD && (
                <div>
                  <DefineMonitor
                    values={values}
                    errors={errors}
                    touched={touched}
                    httpClient={httpClient}
                    location={location}
                    detectorId={this.props.detectorId}
                    notifications={notifications}
                    isDarkMode={isDarkMode}
                  />
                  <EuiSpacer />
                </div>
              )}

              <FieldArray name={'triggerDefinitions'} validateOnChange={true}>
                {(triggerArrayHelpers) => (
                  <ConfigureTriggers
                    edit={edit}
                    triggerArrayHelpers={triggerArrayHelpers}
                    monitor={formikToMonitor(values)}
                    monitorValues={values}
                    setFlyout={this.props.setFlyout}
                    triggers={_.get(formikToMonitor(values), 'triggers', [])}
                    triggerValues={values}
                    isDarkMode={this.props.isDarkMode}
                    httpClient={httpClient}
                    notifications={notifications}
                    notificationService={notificationService}
                    plugins={plugins}
                  />
                )}
              </FieldArray>

              <EuiSpacer />
              <EuiFlexGroup alignItems="center" justifyContent="flexEnd">
                <EuiFlexItem grow={false}>
                  <EuiButtonEmpty onClick={this.onCancel}>Cancel</EuiButtonEmpty>
                </EuiFlexItem>
                <EuiFlexItem grow={false}>
                  <EuiButton fill onClick={handleSubmit} isLoading={isSubmitting}>
                    {edit ? 'Update' : 'Create'}
                  </EuiButton>
                </EuiFlexItem>
              </EuiFlexGroup>
              <SubmitErrorHandler
                errors={errors}
                isSubmitting={isSubmitting}
                isValid={isValid}
                onSubmitError={() =>
                  notifications.toasts.addDanger({
                    title: `Failed to ${edit ? 'update' : 'create'} the monitor`,
                    text: 'Fix all highlighted error(s) before continuing.',
                  })
                }
              />
            </Fragment>
          )}
        </Formik>
      </div>
    );
  }
}
