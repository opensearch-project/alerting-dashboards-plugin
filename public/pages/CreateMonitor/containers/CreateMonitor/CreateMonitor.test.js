/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from 'react';
import { shallow } from 'enzyme';
import { httpServiceMock } from '../../../../../../../src/core/public/mocks';

import CreateMonitor from './CreateMonitor';
import { historyMock, httpClientMock } from '../../../../../test/mocks';
import { FORMIK_INITIAL_VALUES } from './utils/constants';
import AlertingFakes from '../../../../../test/utils/helpers';
import { SEARCH_TYPE } from '../../../../utils/constants';
import { TRIGGER_TYPE } from '../../../CreateTrigger/containers/CreateTrigger/utils/constants';
import { setClient, setNotifications } from '../../../../services';
import { formikToMonitor } from './utils/formikToMonitor';
import coreMock from '../../../../../test/mocks/CoreMock';

const alertingFakes = new AlertingFakes('CreateMonitor random seed');

const setFlyout = jest.fn();
const updateMonitor = jest.fn().mockResolvedValue({ ok: true, id: 'monitor_id' });
const formikBag = {
  setSubmitting: jest.fn(),
  setErrors: jest.fn(),
};
const match = {
  isExact: true,
  params: {},
  path: '/create-monitor',
  url: '/create-monitor',
};
const location = {
  hash: '',
  pathname: '/create-monitor',
  search: '',
  state: undefined,
};
beforeEach(() => {
  jest.clearAllMocks();
});

describe('CreateMonitor', () => {
  const httpClient = httpServiceMock.createStartContract();
  setClient(httpClient);
  setNotifications(coreMock.notifications);
  test('renders', () => {
    const wrapper = shallow(
      <CreateMonitor
        httpClient={httpClientMock}
        history={historyMock}
        setFlyout={setFlyout}
        match={match}
        location={location}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  test('uses AD query Params as initialValues when parameter exists', () => {
    const wrapper = shallow(
      <CreateMonitor
        httpClient={httpClientMock}
        history={historyMock}
        setFlyout={setFlyout}
        match={match}
        location={{ ...location, search: 'searchType=ad&adId=randomId&name=Sample' }}
      />
    );
    expect(wrapper.instance().state.initialValues.name).toBe('Sample-Monitor');
    expect(wrapper.instance().state.initialValues.searchType).toBe(SEARCH_TYPE.AD);
    expect(wrapper.instance().state.initialValues.detectorId).toBe('randomId');
  });

  test('uses monitorToEdit as initialValues when editing', () => {
    const monitor = alertingFakes.randomMonitor();
    const wrapper = shallow(
      <CreateMonitor
        httpClient={httpClientMock}
        edit={true}
        history={historyMock}
        setFlyout={setFlyout}
        updateMonitor={updateMonitor}
        monitorToEdit={monitor}
        match={match}
        location={location}
      />
    );
    expect(wrapper.instance().state.initialValues.name).toBe(monitor.name);
  });

  /**
   * TODO: move these tests to helper.test.js as the new helper.js file has this logic now
  describe('onSubmit', () => {
    test('calls only onUpdate when editing', () => {
      const onCreate = jest.spyOn(CreateMonitor.prototype, 'onCreate');
      const onUpdate = jest.spyOn(CreateMonitor.prototype, 'onUpdate');
      const wrapper = shallow(
        <CreateMonitor
          httpClient={httpClientMock}
          edit={true}
          history={historyMock}
          setFlyout={setFlyout}
          updateMonitor={updateMonitor}
          monitorToEdit={null}
          match={match}
          location={location}
          notifications={coreMock.notifications}
        />
      );
      wrapper.instance().onSubmit(FORMIK_INITIAL_VALUES, formikBag);
      expect(onUpdate).toHaveBeenCalledTimes(1);
      expect(onCreate).not.toHaveBeenCalled();
    });

    test('calls only onCreate when creating', () => {
      const onCreate = jest.spyOn(CreateMonitor.prototype, 'onCreate');
      const onUpdate = jest.spyOn(CreateMonitor.prototype, 'onUpdate');
      httpClientMock.post.mockResolvedValue({ ok: true, resp: { _id: 'monitor_id' } });
      const wrapper = shallow(
        <CreateMonitor
          httpClient={httpClientMock}
          history={historyMock}
          setFlyout={setFlyout}
          match={match}
          location={location}
          notifications={coreMock.notifications}
        />
      );
      wrapper.instance().onSubmit(FORMIK_INITIAL_VALUES, formikBag);
      expect(onCreate).toHaveBeenCalledTimes(1);
      expect(onUpdate).not.toHaveBeenCalled();
    });
  });
  **/

  describe('onCancel', () => {
    test('calls history.goBack if editing', () => {
      const wrapper = shallow(
        <CreateMonitor
          httpClient={httpClientMock}
          edit={true}
          history={historyMock}
          setFlyout={setFlyout}
          updateMonitor={updateMonitor}
          monitorToEdit={null}
          match={match}
          location={location}
        />
      );
      wrapper.instance().onCancel();
      expect(historyMock.goBack).toHaveBeenCalledTimes(1);
      expect(historyMock.push).not.toHaveBeenCalled();
    });

    test('calls history.push when creating', () => {
      const wrapper = shallow(
        <CreateMonitor
          httpClient={httpClientMock}
          history={historyMock}
          setFlyout={setFlyout}
          match={match}
          location={location}
        />
      );
      wrapper.instance().onCancel();
      expect(historyMock.push).toHaveBeenCalledTimes(1);
      expect(historyMock.goBack).not.toHaveBeenCalled();
    });
  });

  describe('onUpdate', () => {
    // Query-level monitor
    test('calls updateMonitor with monitor', () => {
      // const monitor = alertingFakes.randomMonitor();
      // const trigger = alertingFakes.randomTrigger(TRIGGER_TYPE.QUERY_LEVEL);
      // monitor.triggers = [trigger];
      //
      // submitValuesToMonitor(FORMIK_INITIAL_VALUES, )
      const monitor = formikToMonitor(FORMIK_INITIAL_VALUES);
      const wrapper = shallow(
        <CreateMonitor
          httpClient={httpClientMock}
          edit={true}
          history={historyMock}
          setFlyout={setFlyout}
          updateMonitor={updateMonitor}
          monitorToEdit={null}
          match={match}
          location={location}
          notifications={coreMock.notifications}
        />
      );
      wrapper.instance().onSubmit(FORMIK_INITIAL_VALUES, formikBag);
      expect(updateMonitor).toHaveBeenCalledTimes(1);
      expect(updateMonitor).toHaveBeenCalledWith(monitor);
    });

    test('logs error when updateMonitor rejects', async () => {
      const error = jest.spyOn(global.console, 'error');
      updateMonitor.mockRejectedValue(new Error('updateMonitor error'));
      const wrapper = shallow(
        <CreateMonitor
          httpClient={httpClientMock}
          edit={true}
          history={historyMock}
          setFlyout={setFlyout}
          updateMonitor={updateMonitor}
          monitorToEdit={null}
          match={match}
          location={location}
          notifications={coreMock.notifications}
        />
      );
      await wrapper.instance().onSubmit(FORMIK_INITIAL_VALUES, formikBag);
      expect(error).toHaveBeenCalled();
    });

    test('logs resp when ok:false', async () => {
      const log = jest.spyOn(global.console, 'log');
      updateMonitor.mockResolvedValue({ ok: false, resp: 'test' });
      const wrapper = shallow(
        <CreateMonitor
          httpClient={httpClientMock}
          edit={true}
          history={historyMock}
          setFlyout={setFlyout}
          updateMonitor={updateMonitor}
          monitorToEdit={null}
          match={match}
          location={location}
          notifications={coreMock.notifications}
        />
      );
      await wrapper.instance().onSubmit(FORMIK_INITIAL_VALUES, formikBag);
      await new Promise((r) => setTimeout(r, 100));
      expect(log).toHaveBeenCalled();
      expect(log).toHaveBeenCalledWith('Failed to update:', { ok: false, resp: 'test' });
    });
  });

  describe('onCreate', () => {
    test('calls post with monitor', () => {
      const monitor = formikToMonitor(FORMIK_INITIAL_VALUES);
      httpClientMock.post.mockResolvedValue({ ok: true, resp: { _id: 'id' } });
      const wrapper = shallow(
        <CreateMonitor
          httpClient={httpClientMock}
          history={historyMock}
          setFlyout={setFlyout}
          match={match}
          location={location}
          notifications={coreMock.notifications}
        />
      );
      wrapper.instance().onSubmit(FORMIK_INITIAL_VALUES, formikBag);
      expect(httpClientMock.post).toHaveBeenCalledTimes(1);
      expect(httpClientMock.post).toHaveBeenCalledWith('../api/alerting/monitors', {
        body: JSON.stringify(monitor),
      });
    });

    test('logs error when updateMonitor rejects', async () => {
      const error = jest.spyOn(global.console, 'error');
      httpClientMock.post.mockRejectedValue(new Error('onCreate error'));
      const wrapper = shallow(
        <CreateMonitor
          httpClient={httpClientMock}
          history={historyMock}
          setFlyout={setFlyout}
          match={match}
          location={location}
          notifications={coreMock.notifications}
        />
      );
      await wrapper.instance().onSubmit(FORMIK_INITIAL_VALUES, formikBag);
      expect(error).toHaveBeenCalled();
    });

    test('logs resp when ok:false', async () => {
      const log = jest.spyOn(global.console, 'log');
      httpClientMock.post.mockResolvedValue({ ok: false, resp: 'test' });
      const wrapper = shallow(
        <CreateMonitor
          httpClient={httpClientMock}
          history={historyMock}
          setFlyout={setFlyout}
          match={match}
          location={location}
          notifications={coreMock.notifications}
        />
      );
      await wrapper.instance().onSubmit(FORMIK_INITIAL_VALUES, formikBag);
      await new Promise((r) => setTimeout(r, 100));
      expect(log).toHaveBeenCalled();
      expect(log).toHaveBeenCalledWith('Failed to create:', { ok: false, resp: 'test' });
    });
  });
});
