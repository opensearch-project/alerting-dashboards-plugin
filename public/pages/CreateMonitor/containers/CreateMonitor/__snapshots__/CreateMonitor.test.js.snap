// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateMonitor renders 1`] = `
<div
  style={
    Object {
      "padding": "25px 50px",
    }
  }
>
  <Formik
    initialValues={
      Object {
        "adResultIndex": undefined,
        "aggregationType": "count",
        "aggregations": Array [],
        "associatedMonitors": Object {
          "sequence": Object {
            "delegates": Array [],
          },
        },
        "associatedMonitorsEditor": "",
        "associatedMonitorsList": Array [],
        "bucketUnitOfTime": "h",
        "bucketValue": 1,
        "clusterNames": Array [],
        "cronExpression": "0 */1 * * *",
        "daily": 0,
        "description": "",
        "detectorId": "",
        "disabled": false,
        "fieldName": Array [],
        "filters": Array [],
        "frequency": "interval",
        "groupBy": Array [],
        "groupByField": Array [
          Object {
            "label": "",
          },
        ],
        "groupedOverFieldName": "bytes",
        "groupedOverTop": 5,
        "index": Array [],
        "monitor_type": "query_level_monitor",
        "monthly": Object {
          "day": 1,
          "type": "day",
        },
        "name": "",
        "overDocuments": "all documents",
        "period": Object {
          "interval": 1,
          "unit": "MINUTES",
        },
        "preventVisualEditor": false,
        "queries": Array [],
        "query": "{
    \\"size\\": 0,
    \\"query\\": {
        \\"match_all\\": {}
    }
}",
        "searchType": "graph",
        "timeField": "",
        "timezone": Array [],
        "uri": Object {
          "api_type": "",
          "clusters": Array [],
          "path": "",
          "path_params": "",
          "url": "",
        },
        "weekly": Object {
          "fri": false,
          "mon": false,
          "sat": false,
          "sun": false,
          "thur": false,
          "tue": false,
          "wed": false,
        },
      }
    }
    onSubmit={[Function]}
    validateOnChange={false}
  >
    <Component />
  </Formik>
</div>
`;
